examples/step-58/step-58.cc

```diff
@@ -0,0 +1,673 @@
+  // first and last part of each time step.
+  //
+  // To this end, recall that for the first half step, we needed to
+  // compute
+  // $\psi^{(n,1)} = e^{-i\kappa|\psi^{(n,0)}|^2 \tfrac
+  //  12\Delta t} \; \psi^{(n,0)}$. Here, $\psi^{(n,0)}=\psi^{(n)}$ and
+  //  $\psi^{(n,1)}$
+  // are functions of space and correspond to the output of the previous
+  // complete time step and the result of the first of the three part steps,
+  // respectively. A corresponding solution must be computed for the third
+  // of the part steps, i.e.
+  // $\psi^{(n,3)} = e^{-i\kappa|\psi^{(n,2)}|^2 \tfrac
+  //  12\Delta t} \; \psi^{(n,2)}$, where $\psi^{(n,3)}=\psi^{(n+1)}$ is
+  // the result of the time step as a whole, and its input $\psi^{(n,2)}$ is
+  // the result of the spatial step of the Strange splitting.
+  //
+  // An important realization is that while $\psi^{(n,0)}(\mathbf x)$ may be a
+  // finite element function (i.e., is piecewise polynomial), this may not
+  // necessarily be the case for the "rotated" function in which we have updated
+  // the phase using the exponential factor (recall that the amplitude of that
+  // function remains constant as part of that step). In other words, we could
+  // *compute* $\psi^{(n,1)}(\mathbf x)$ at every point $\mathbf x\in\Omega$,
+  // but we can't represent it on a mesh because it is not a piecewise
+  // polynomial function. The best we can do in a discrete setting is to compute
+  // a projection or interpolation. In other words, we can compute
+  // $\psi_h^{(n,1)}(\mathbf x) = \Pi_h
+  //     \left(e^{-i\kappa|\psi_h^{(n,0)}(\mathbf x)|^2 \tfrac 12\Delta t}
+  //     \; \psi_h^{(n,0)}(\mathbf x) \right)$ where $\Pi_h$ is a projection or
+  // interpolation operator. The situation is particularly simple if we
+  // choose the interpolation: Then, all we need to compute is the value of
+  // the right hand side *at the node points* and use these as nodal
+  // values for the vector $\Psi^{(n,1)}$ of degrees of freedom. This is
+  // easily done because evaluating the right hand side at node points
+  // for a Lagrange finite element as used here requires us to only
+  // look at a single (complex-valued) entry of the node vector. In other
+  // words, what we need to do is to compute
+  // $\Psi^{(n,1)}_j = e^{-i\kappa|\Psi^{(n,0)}_j|^2 \tfrac
+  //  12\Delta t} \; \Psi^{(n,0)}_j$ where $j$ loops over all of the entries
+  // of our solution vector. This is what the function below does -- in fact,
+  // it doesn't even use separate vectors for $\Psi^{(n,0)}$ and $\Psi^{(n,1)}$,
+  // but just updates the same vector as appropriate.
+  template <int dim>
+  void NonlinearSchroedingerEquation<dim>::do_half_phase_step()
+  {
+    for (auto &value : solution)
+      {
+        const std::complex<double> i         = {0, 1};
+        const double               magnitude = std::abs(value);
+
+        value = std::exp(-i * kappa * magnitude * magnitude * (time_step / 2)) *
+                value;
```

Comment:
Couldn't you write this as `value *= std::exp(...);`? Then its really clear that it represents an update.

Reply:
I could, but I'd like to keep it this way because it matches the formulas: `psi^{n+1} = exp(...) psi^n`. Using `*=` obscures the fact that we actually move from one stage to another.